### PR TITLE
refactor geosearch control to port search logic into an external class

### PR DESCRIPTION
**Pull request by jgravois**
anyone game for giving a little code review on my latest attempt to restructure the logic of `L.esri.Geocoding.geosearch()` so that it will be possible for folks to utilize `<input>` tags outside the map as well?

at this point i'm adding a `_geosearchClass` property to the control and a `_control` property to the class so that the two can share logic

to do:
- [x] `Classes/GeosearchClass` probably needs a rename
- [x] the last of the DOM manipulation within `GeosearchClass._suggest()` needs to be moved to the control
- [x] more thorough testing of the original control.  i've been using `debug/sample.html`, but we need to test with custom constructor options and providers as well.
- [x] deciding which methods really need to be private.
- [x] figure out why the reverseGeocode stuff that has already been added to `master` is showing up in this PR too (something went wrong when i pulled from upstream obviously)
- [x] write another sample that puts the map on one half of the page and the input on the other
- [x] make a few more tweaks to get the last of my bootstrap.css hacks out of the geosearchInput source and into the sample itself.
- [x] figure out why you can't tab/arrow through suggestions
- [x] if `geosearchInput` isnt going to apply any .css automatically, developers need a mechanism to apply their own classes to the wrapper and suggestion divs.
- [x] figure out why events fired by `geosearchCore` (like `results`) don't 'bubble up' to the control (use the [source](https://github.com/Leaflet/Leaflet/blob/master/src/core/Events.js#L198-L203) luke)
- [x] fix the fact that only results from one provider at a time are being displayed
- [x] make sure `geosearchInput` adds attribution to the map (#108)
- [x] bootstrap `dropdown-menu` only allows a single `<ul>`.  ~~either refactor more substantially for that case or~~ just doc limitation of a single provider

### CODE REVIEW

**Comment by ngoldman** [comment]
Looks pretty decent to me at first glance.

File: src/Controls/Geosearch.js [hunk 1]
```diff
@@ -386,19 +229,30 @@ export var Geosearch = L.Control.extend({
...
-    L.DomEvent.addListener(this._suggestions, 'mouseout', function (e) {
+    L.DomEvent.addListener(this._geosearchClass._suggestions, 'mouseout', function (e) {
       if (!map.scrollWheelZoom.enabled() && map.options.scrollWheelZoom) {
         map.scrollWheelZoom.enable();
       }
     });
 
+    // using 'this._control' feels like a hack
+    this._geosearchClass.on('no-active-requests', function (e) {
+      L.DomUtil.removeClass(this._control._input, 'geocoder-control-loading');
```
Comment (jgravois) [inline src/Controls/Geosearch.js]: seems weird to me to have to use `this._control._input` here...

Comment (ngoldman) [inline src/Controls/Geosearch.js]: yeah, accessing private variables is usually a bad idea as they can disappear without notice.

Comment (jgravois) [inline src/Controls/Geosearch.js]: i was expecting `this._input` to resolve, but i guess when i moved to using an event listener from the other class, it started passing its own context back.

**Comment by jgravois** [comment]
caught the fact that my new sample control failed to pass along the `magicKey`.  i still need to look at how to resolve the fact that too many suggest requests are being triggered and make css less hardcoded.

**Comment by jgravois** [comment]
@patrickarlt 
  1. are you okay with me adding an additional class/control/util for off map geosearch to this repo or would you rather i created a seperate plugin?
  2. either way, can you think of a better name than 'geosearchOffMap'?
  3. can you suggest a better name than `GeosearchClass` for the object i broke out from your control?
(this PR really has me questioning my creativity)

**Comment by patrickarlt** [comment]
@jgravois for the class name how about `geosearchInput`? It looks like `GeosearchClass` is required both my the control and the input so how about just calling `GeosearchCore`?

File: src/Controls/GeosearchOffMap.js [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+    return attribution;
+  },
+
+  setup: function (options) {
+    this._map = options.map;
+    this._input = document.getElementById(options.inputTag)
+    L.DomUtil.addClass(this._input, 'the-input');
+
+    // more neckbeardy if i can stop using jquery
+    this._wrapper = $( ".the-input" ).wrap(function() {
```
Comment (patrickarlt) [inline src/Controls/GeosearchOffMap.js]: @jgravois we need to drop the jQuery requirement here.

File: src/Controls/GeosearchOffMap.js [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+      return "<div id= 'address-foobaz' class='input-group' style='width:100%'></div>";
+    });
+
+    this._input.title = this.options.title;
+
+    // more neckbeardy if i can stop using jquery
+    $( "<div id='sugg' class='suggestions-group'></div>" ).insertAfter( ".the-input" );
```
Comment (patrickarlt) [inline src/Controls/GeosearchOffMap.js]: Need to drop jQuery.

File: src/Controls/GeosearchOffMap.js [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+  },
+
+  setup: function (options) {
+    this._map = options.map;
+    this._input = document.getElementById(options.inputTag)
+    L.DomUtil.addClass(this._input, 'the-input');
+
+    // more neckbeardy if i can stop using jquery
+    this._wrapper = $( ".the-input" ).wrap(function() {
+      return "<div id= 'address-foobaz' class='input-group' style='width:100%'></div>";
```
Comment (patrickarlt) [inline src/Controls/GeosearchOffMap.js]: `address-foobaz` as the ID is a little odd. I'm assuming you wanted a unique id? The best way to do that is to keep a private counter of how many times this has been called and generate an ID with that number.

**Comment by patrickarlt** [comment]
@jgravois made some inline comments. You also need to get this passing the linter. Once that gets done i'll do a more through review.

The other alternative is to abandon this approach and just make an entirely separate repo based on something like web components, which might make more sense.

**Comment by jgravois** [comment]
i'm totally into trying to help do something using web components, but based on the number of requests i've gotten for the functionality, i'd like to merge this work in in the meantime so that devs have something available now.

* i've renamed the classes based on your good suggestions
* i've also get rid of the last remaining jQuery references
* code is now linted
* all tests are passing

to do:
- [ ] write another sample that puts the map on one half of the page and the input on the other
- [ ] makle a few more tweaks to get the last of my bootstrap.css hacks out of the geosearchInput source and into the sample itself.

i'll let you know when that stuff is done so you can take one more good look at the code.

thanks **so much** for the code review!

File: src/Classes/GeosearchInput.js [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+        list = L.DomUtil.create('ul', this.options.styling.suggestionsList || null, this._suggestions);
+        list.style.display = 'block';
+      }
+      var suggestionItem = L.DomUtil.create('li', 'geocoder-control-suggestion', list);
+      // need to wrap in an <a> tag to get selection CSS and good spacing
+      suggestionItem.innerHTML = '<a>' + suggestion.text + '</a>';
+
+      // this means we need to dig into the child node to stuff in our properties.
+      // not ideal.
+      suggestionItem.childNodes[0].provider = suggestion.provider;
```
Comment (jgravois) [inline src/Classes/GeosearchInput.js]: i *know* theres got to be a better solution than this.

**Comment by jgravois** [comment]
perfect has been the enemy of progress for too long so i decided to limit the scope of this PR to breaking up the logic of `L.esri.Geocoding.geosearch()` into a control which handles UI and an underlying class that handles the geocoding and suggestions.  this gets us closer to implementing a fix for #32.

all tests are passing and my own manual testing confirms that the changes have zero impact on the control that is actually surfaced in the API.  all said, code review prior to merge would be sincerely appreciated 

cc/ @Esri/leaflet-contributors

**Pull request merged**